fish-rust/src/builtins/string.rs

```diff
@@ -0,0 +1,493 @@
+use std::borrow::Cow;
+use std::fs::File;
+use std::io::{BufRead, BufReader, Read};
+use std::os::fd::FromRawFd;
+
+use crate::builtins::shared::{
+    builtin_missing_argument, builtin_print_error_trailer, builtin_print_help, io_streams_t,
+    BUILTIN_ERR_INVALID_SUBCMD, BUILTIN_ERR_MISSING_SUBCMD, BUILTIN_ERR_NOT_NUMBER,
+    BUILTIN_ERR_TOO_MANY_ARGUMENTS, BUILTIN_ERR_UNKNOWN, STATUS_CMD_OK, STATUS_INVALID_ARGS,
+};
+use crate::common::str2wcstring;
+use crate::ffi::parser_t;
+use crate::wchar::{wstr, WString, L};
+use crate::wchar_ext::WExt;
+use crate::wcstringutil::fish_wcwidth_visible;
+use crate::wgetopt::{wgetopter_t, woption};
+use crate::wutil::{sprintf, wgettext_fmt};
+use libc::c_int;
+
+mod collect;
+mod escape;
+mod join;
+mod length;
+mod r#match;
+mod pad;
+mod repeat;
+mod replace;
+mod shorten;
+mod split;
+mod sub;
+mod transform;
+mod trim;
+mod unescape;
+
+macro_rules! string_error {
+    (
+    $streams:expr,
+    $string:expr
+    $(, $args:expr)+
+    $(,)?
+    ) => {
+        $streams.err.append(L!("string "));
+        $streams.err.append(wgettext_fmt!($string, $($args),*));
+    };
+}
+pub(self) use string_error;
+
+fn string_unknown_option(
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+    subcmd: &wstr,
+    opt: &wstr,
+) {
+    string_error!(streams, BUILTIN_ERR_UNKNOWN, subcmd, opt);
+    builtin_print_error_trailer(parser, streams, L!("string"));
+}
+
+trait StringSubCommand<'args> {
+    const SHORT_OPTIONS: &'static wstr;
+    const LONG_OPTIONS: &'static [woption<'static>];
+
+    /// Parse and store option specificed by the associated short or long option.
+    fn parse_opt(&mut self, optarg: Option<&'args wstr>, c: char) -> Result<(), StringError>;
+
+    fn parse_opts(
+        &mut self,
+        args: &mut [&'args wstr],
+        parser: &mut parser_t,
+        streams: &mut io_streams_t,
+    ) -> Result<usize, Option<c_int>> {
+        let cmd = args[0];
+        let mut args_read = Vec::with_capacity(args.len());
+        args_read.extend_from_slice(args);
+
+        let mut w = wgetopter_t::new(Self::SHORT_OPTIONS, Self::LONG_OPTIONS, args);
```

Comment:
The copy is ugly.
wgetopter_t wants exclusive access to args (because it will rearrange args so positionals are next to each other).
How about giving wgetopt accessor methods like `w.cmd()` and `w.argv()`?

There's another issue: `args: &mut [&wstr]` has the wrong type (for all builtins).
In practice it is always underpinned by a `&mut Vec<WString>` we should change
it to `args: &mut [WString]` to avoid the need to allocate a slice-of-slices;

Here's some of that change I made a while ago:

<details>

```diff
diff --git a/fish-rust/src/wgetopt.rs b/fish-rust/src/wgetopt.rs
index bc4224c98..0ff56323f 100644
--- a/fish-rust/src/wgetopt.rs
+++ b/fish-rust/src/wgetopt.rs
@@ -23,7 +23,9 @@ License along with the GNU C Library; see the file COPYING.LIB.  If
 not, write to the Free Software Foundation, Inc., 675 Mass Ave,
 Cambridge, MA 02139, USA.  */

-use crate::wchar::{wstr, WExt, L};
+use std::ops;
+
+use crate::wchar::{wstr, WExt, WString, L};

 /// Describe how to deal with options that follow non-option ARGV-elements.
 ///
@@ -64,14 +66,14 @@ fn empty_wstr() -> &'static wstr {
     Default::default()
 }

-pub struct wgetopter_t<'opts, 'args, 'argarray> {
+pub struct wgetopter_t<'opts, 'argarray> {
     /// Argv.
-    argv: &'argarray mut [&'args wstr],
+    argv: &'argarray mut [WString],

     /// For communication from `getopt` to the caller. When `getopt` finds an option that takes an
     /// argument, the argument value is returned here. Also, when `ordering` is RETURN_IN_ORDER, each
     /// non-option ARGV-element is returned here.
-    pub woptarg: Option<&'args wstr>,
+    woptarg_idx: Option<(usize, usize)>,

     shortopts: &'opts wstr,
     longopts: &'opts [woption<'opts>],
@@ -80,7 +82,13 @@ pub struct wgetopter_t<'opts, 'args, 'argarray> {
     /// returned was found. This allows us to pick up the scan where we left off.
     ///
     /// If this is empty, it means resume the scan by advancing to the next ARGV-element.
+    nextchar_slice: ops::Range<usize>,

     /// Index in ARGV of the next element to be scanned. This is used for communication to and from
     /// the caller and for communication between successive calls to `getopt`.
@@ -153,11 +161,11 @@ pub const fn wopt(name: &wstr, has_arg: woption_argument_t, val: char) -> woptio
     woption { name, has_arg, val }
 }

-impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
+impl<'opts, 'argarray> wgetopter_t<'opts, 'argarray> {
     pub fn new(
         shortopts: &'opts wstr,
         longopts: &'opts [woption],
-        argv: &'argarray mut [&'args wstr],
+        argv: &'argarray mut [WString],
     ) -> Self {
         return wgetopter_t {
             woptopt: '?',
@@ -168,9 +176,9 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
             initialized: false,
             last_nonopt: 0,
             missing_arg_return_colon: false,
-            nextchar: Default::default(),
+            nextchar_slice: 0..0,
             ordering: Ordering::PERMUTE,
-            woptarg: None,
+            woptarg_idx: None,
             woptind: 0,
         };
     }
@@ -185,6 +193,23 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
         return self._wgetopt_internal(opt_index, false);
     }

+    pub fn woptarg(&self) -> Option<&wstr> {
+        self.woptarg_idx
+            .map(|(woptind, start)| &self.argv[woptind][start..])
+    }
+
+    pub fn cmd(&self) -> &wstr {
+        &self.argv[0]
+    }
+
+    pub fn argv(&self) -> &[WString] {
+        self.argv
+    }
+
+    pub fn nextchar(&self) -> &wstr {
+        &self.argv[self.woptind][self.nextchar_slice.clone()]
+    }
+
     /// \return the number of arguments.
     fn argc(&self) -> usize {
         return self.argv.len();
@@ -241,7 +266,7 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
         self.first_nonopt = 1;
         self.last_nonopt = 1;
         self.woptind = 1;
-        self.nextchar = empty_wstr();
+        self.nextchar_slice = 0..0;

         let mut optstring = self.shortopts;

@@ -321,7 +346,7 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
             if self.ordering == Ordering::REQUIRE_ORDER {
                 return None;
             }
-            self.woptarg = Some(self.argv[self.woptind]);
+            self.woptarg_idx = Some((self.woptind, 0));
             self.woptind += 1;
             return Some(char::from(1));
         }
@@ -332,15 +357,15 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
         } else {
             1
         };
-        self.nextchar = self.argv[self.woptind][skip..].into();
+        self.nextchar_slice = skip..(self.argv[self.woptind].len());
         return Some(char::from(0));
     }

     /// Check for a matching short opt.
     fn _handle_short_opt(&mut self) -> char {
         // Look at and handle the next short option-character.
-        let mut c = self.nextchar.char_at(0);
-        self.nextchar = &self.nextchar[1..];
+        let mut c = self.nextchar().char_at(0);
+        self.nextchar_slice = self.nextchar_slice.start + 1..self.nextchar_slice.end;

         let temp = match self.shortopts.chars().position(|sc| sc == c) {
             Some(pos) => &self.shortopts[pos..],
@@ -348,14 +373,14 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
         };

         // Increment `woptind' when we start to process its last character.
-        if self.nextchar.is_empty() {
+        if self.nextchar().is_empty() {
             self.woptind += 1;
         }

         if temp.is_empty() || c == ':' {
             self.woptopt = c;

-            if !self.nextchar.is_empty() {
+            if !self.nextchar().is_empty() {
                 self.woptind += 1;
             }
             return '?';
@@ -367,17 +392,17 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {

         if temp.char_at(2) == ':' {
             // This is an option that accepts an argument optionally.
-            if !self.nextchar.is_empty() {
-                self.woptarg = Some(self.nextchar);
+            if !self.nextchar().is_empty() {
+                self.woptarg_idx = Some((self.woptind, 0));
                 self.woptind += 1;
             } else {
-                self.woptarg = None;
+                self.woptarg_idx = None;
             }
-            self.nextchar = empty_wstr();
+            self.nextchar_slice = 0..0;
         } else {
             // This is an option that requires an argument.
-            if !self.nextchar.is_empty() {
-                self.woptarg = Some(self.nextchar);
+            if !self.nextchar().is_empty() {
+                self.woptarg_idx = Some((self.woptind, 0));
                 // If we end this ARGV-element by taking the rest as an arg, we must advance to
                 // the next element now.
                 self.woptind += 1;
@@ -391,10 +416,10 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
             } else {
                 // We already incremented `woptind' once; increment it again when taking next
                 // ARGV-elt as argument.
-                self.woptarg = Some(self.argv[self.woptind]);
+                self.woptarg_idx = Some((self.woptind, 0));
                 self.woptind += 1;
             }
-            self.nextchar = empty_wstr();
+            self.nextchar_slice = 0..0;
         }

         return c;
@@ -409,21 +434,23 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
         retval: &mut char,
     ) {
         self.woptind += 1;
-        assert!(self.nextchar.char_at(nameend) == '\0' || self.nextchar.char_at(nameend) == '=');
-        if self.nextchar.char_at(nameend) == '=' {
+        assert!(
+            self.nextchar().char_at(nameend) == '\0' || self.nextchar().char_at(nameend) == '='
+        );
+        if self.nextchar().char_at(nameend) == '=' {
             if pfound.has_arg != woption_argument_t::no_argument {
-                self.woptarg = Some(self.nextchar[(nameend + 1)..].into());
+                self.woptarg_idx = Some((self.woptind, nameend + 1));
             } else {
-                self.nextchar = empty_wstr();
+                self.nextchar_slice = 0..0;
                 *retval = '?';
                 return;
             }
         } else if pfound.has_arg == woption_argument_t::required_argument {
             if self.woptind < self.argc() {
-                self.woptarg = Some(self.argv[self.woptind]);
+                self.woptarg_idx = Some((self.woptind, 0));
                 self.woptind += 1;
             } else {
-                self.nextchar = empty_wstr();
+                self.nextchar_slice = 0..0;
                 *retval = if self.missing_arg_return_colon {
                     ':'
                 } else {
@@ -433,7 +460,7 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
             }
         }

-        self.nextchar = empty_wstr();
+        self.nextchar_slice = 0..0;
         *longind = option_index;
         *retval = pfound.val;
     }
@@ -451,7 +478,7 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
         // Test all long options for either exact match or abbreviated matches.
         for (option_index, p) in self.longopts.iter().enumerate() {
             // Check if current option is prefix of long opt
-            if p.name.starts_with(&self.nextchar[..nameend]) {
+            if p.name.starts_with(&self.nextchar()[..nameend]) {
                 if nameend == p.name.len() {
                     // The current option is exact match of this long option
                     pfound = Some(*p);
@@ -483,14 +510,14 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
         let mut indfound: usize = 0;

         let mut nameend = 0;
-        while self.nextchar.char_at(nameend) != '\0' && self.nextchar.char_at(nameend) != '=' {
+        while self.nextchar().char_at(nameend) != '\0' && self.nextchar().char_at(nameend) != '=' {
             nameend += 1;
         }

         let pfound = self._find_matching_long_opt(nameend, &mut exact, &mut ambig, &mut indfound);

         if ambig && !exact {
-            self.nextchar = empty_wstr();
+            self.nextchar_slice = 0..0;
             self.woptind += 1;
             *retval = '?';
             return true;
@@ -509,9 +536,9 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
             || !self
                 .shortopts
                 .as_char_slice()
-                .contains(&self.nextchar.char_at(0))
+                .contains(&self.nextchar().char_at(0))
         {
-            self.nextchar = empty_wstr();
+            self.nextchar_slice = 0..0;
             self.woptind += 1;
             *retval = '?';
             return true;
@@ -541,7 +568,7 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
     /// If a char in OPTSTRING is followed by a colon, that means it wants an arg, so the following text
     /// in the same ARGV-element, or the text of the following ARGV-element, is returned in `optarg`.
     /// Two colons mean an option that wants an optional arg; if there is text in the current
-    /// ARGV-element, it is returned in `w.woptarg`, otherwise `w.woptarg` is set to zero.
+    /// ARGV-element, it is returned in `w.woptarg()`, otherwise `w.woptarg()` is set to zero.
     ///
     /// If OPTSTRING starts with `-` or `+', it requests different methods of handling the non-option
     /// ARGV-elements. See the comments about RETURN_IN_ORDER and REQUIRE_ORDER, above.
@@ -563,9 +590,9 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
         if !self.initialized {
             self._wgetopt_initialize();
         }
-        self.woptarg = None;
+        self.woptarg_idx = None;

-        if self.nextchar.is_empty() {
+        if self.nextchar().is_empty() {
             let narg = self._advance_to_next_argv();
             if narg != Some(char::from(0)) {
                 return narg;
@@ -586,7 +613,7 @@ impl<'opts, 'args, 'argarray> wgetopter_t<'opts, 'args, 'argarray> {
         //
         // This distinction seems to be the most useful approach.
         if !self.longopts.is_empty() && self.woptind < self.argc() {
-            let arg = self.argv[self.woptind];
+            let arg = &self.argv[self.woptind];

             #[allow(clippy::if_same_then_else)]
             #[allow(clippy::needless_bool)]
```

</details>

Reply:
I agree the cloning is ugly, but this is independent of this PR and present in other subcommands as well, I would prefer to not batch that into this already quite large change